src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java

```diff
@@ -538,347 +511,128 @@ private ObjectMetadata syncPutObject(
-                throw new PidRefsFileExistsException(errMsg);
-
-            } else if (Files.exists(absCidRefsPath)) {
-                // Only update cid refs file if pid is not in the file
-                boolean pidFoundInCidRefFiles = isPidInCidRefsFile(pid, absCidRefsPath);
-                if (!pidFoundInCidRefFiles) {
-                    updateCidRefsFiles(pid, absCidRefsPath);
-                }
-                // Get the pid refs file
-                File pidRefsTmpFile = writePidRefsFile(cid);
-                File absPathPidRefsFile = absPidRefsPath.toFile();
-                move(pidRefsTmpFile, absPathPidRefsFile, "refs");
-                // Verify tagging process, this throws exceptions if there's an issue
-                verifyHashStoreRefsFiles(pid, cid, absPidRefsPath, absCidRefsPath);
-
-                logFileHashStore.info(
-                    "FileHashStore.tagObject - Object with cid: " + cid
-                        + " has been updated and tagged successfully with pid: " + pid
-                );
+            synchronizeObjectLockedCids(cid);
+            synchronizeReferenceLockedPids(pid);
+            storeHashStoreRefsFiles(pid, cid);
 
-            } else {
-                // Get pid and cid refs files 
-                File pidRefsTmpFile = writePidRefsFile(cid);
-                File cidRefsTmpFile = writeCidRefsFile(pid);
-                // Move refs files to permanent location
-                File absPathPidRefsFile = absPidRefsPath.toFile();
-                File absPathCidRefsFile = absCidRefsPath.toFile();
-                move(pidRefsTmpFile, absPathPidRefsFile, "refs");
-                move(cidRefsTmpFile, absPathCidRefsFile, "refs");
-                // Verify tagging process, this throws exceptions if there's an issue
-                verifyHashStoreRefsFiles(pid, cid, absPidRefsPath, absCidRefsPath);
-
-                logFileHashStore.info(
-                    "FileHashStore.tagObject - Object with cid: " + cid
-                        + " has been tagged successfully with pid: " + pid
-                );
-            }
+        } catch (HashStoreRefsAlreadyExistException hsrfae) {
+            // This exception is thrown when the pid and cid are already tagged appropriately
+            String errMsg =
+                "HashStore refs files already exist for pid " + pid + " and cid: " + cid;
+            throw new HashStoreRefsAlreadyExistException(errMsg);
+
+        } catch (PidRefsFileExistsException prfe) {
+            String errMsg = "pid: " + pid + " already references another cid."
+                + " A pid can only reference one cid.";
+            throw new PidRefsFileExistsException(errMsg);
+
+        } catch (Exception e) {
+            // Revert the process for all other exceptions
+            // We must first release the cid and pid since 'unTagObject' is synchronized
+            // If not, we will run into a deadlock.
+            releaseObjectLockedCids(cid);
```

Comment:
The code releases the locks in the finally block, any reasons why we should release them in the catch block again?

Reply:
In the scenario that an exception is encountered, the `pid` and `cid`s related to the process could be in a various states. The sync values are no longer relevant, and `unTagObject` attempts to determine where everything is currently at, before untagging the data object.

Since `unTagObject` is synchronized and follows roughly the same process a delete operation, we must release these values first - otherwise we will run into a dead lock as it tries to safely untag a data object.